### PR TITLE
Iep 596 add new post endpoint for liq03 api

### PR DESCRIPTION
**Pull request by adeharrison-ch**

### CODE REVIEW

File: handlers/progress_report_resource.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		transactionID, validTransactionId := utils.HandleTransactionIdExistsValidation(w, req, utils.GetTransactionIDFromVars(mux.Vars(req)))
+		if !validTransactionId {
+			return
+		}
+
+		log.InfoR(req, fmt.Sprintf("start POST request for submit progress report with transaction id: %s", transactionID))
+
+		isTransactionClosed, err, httpStatus := service.CheckIfTransactionClosed(transactionID, req)
+		err, validTransactionNotClosed := utils.HandleTransactionNotClosedValidation(w, req, transactionID, isTransactionClosed, err, httpStatus)
```
Comment (Jideboris) [inline handlers/progress_report_resource.go:28]: We need to handle this err

File: handlers/progress_report_resource.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		if !validTransactionNotClosed {
+			return
```
Comment (Jideboris) [inline handlers/progress_report_resource.go:30]: There should be log here to inform what happned

File: handlers/progress_report_resource.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		}
+
+		//todo replace using generics when GO version 1.18+
+		var request models.ProgressReport
+		err = json.NewDecoder(req.Body).Decode(&request)
```
Comment (Jideboris) [inline handlers/progress_report_resource.go:35]: if err at this point what happens??

File: handlers/progress_report_resource.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+		// Request body failed to get decoded
+		if !utils.HandleBodyDecodedValidation(w, req, transactionID, err) {
+			return
+		}
+
+		//todo replace using generics when GO version 1.18+
+		progressReportDao := transformers.ProgressReportResourceRequestToDB(&request)
```
Comment (Jideboris) [inline handlers/progress_report_resource.go:43]: Check if not nil should be considered I think.

File: transformers/progress_report_resource.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package transformers
+
+import (
+	"github.com/companieshouse/insolvency-api/models"
+	"github.com/companieshouse/insolvency-api/utils"
+)
+
+// ProgressReportResourceRequestToDB transforms a progress report request to a dao model
+func ProgressReportResourceRequestToDB(req *models.ProgressReport) *models.ProgressReportResourceDao {
+	etag, err := utils.GenerateEtag()
```
Comment (Jideboris) [inline transformers/progress_report_resource.go:10]: Check if err is not nil and log is important here

Comment (Jideboris) [inline transformers/progress_report_resource.go:10]: etag expected value worth checking as well so it does not panic along the line

File: utils/common_validation.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
+package utils
```
Comment (Jideboris) [inline utils/common_validation.go:1]: I think utils should be injected as interface to service struct for mocking and testing sakes.

File: utils/common_validation_test.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package utils
```
Comment (Jideboris) [inline utils/common_validation_test.go:1]: I think utils should be injected as interface to service struct for mocking and testing sakes.

File: handlers/progress_report_resource.go (hunk 1 above)
Comment (adeharrison-ch) [reply handlers/progress_report_resource.go:28]: changed to _ as not required

File: handlers/progress_report_resource.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+			return
+		}
+
+		//todo replace using generics when GO version 1.18+
+		var request models.ProgressReport
+		err = json.NewDecoder(req.Body).Decode(&request)
+
+		// Request body failed to get decoded
+		if !utils.HandleBodyDecodedValidation(w, req, transactionID, err) {
+			return
```
[the remark this answers is not in the document]
Comment (adeharrison-ch) [reply handlers/progress_report_resource.go:39]: other than logged within HandleBodyDecodedValidation()?

File: transformers/progress_report_resource.go (hunk 5 above)
Comment (adeharrison-ch) [reply transformers/progress_report_resource.go:10]: done

File: handlers/progress_report_resource.go (hunk 4 above)
Comment (adeharrison-ch) [reply handlers/progress_report_resource.go:43]: not sure how to force failure in a test due to strong typing?

File: utils/common_validation_test.go (hunk 7 above)
Comment (adeharrison-ch) [reply utils/common_validation_test.go:1]: Can we accept this as is until IEP-634 is ready to go as no idea yet how to implement that

File: utils/common_validation.go (hunk 6 above)
Comment (adeharrison-ch) [reply utils/common_validation.go:1]: Can we accept this as is until IEP-634 is ready to go as no idea yet how to implement that

**Review by Jideboris — APPROVED** [review]

**Pull request merged**